### DISCUSSION TITLE
Is there a way to setup input.conf to spit out the target's MaxCLL/Peak Luminance?

### DISCUSSION THREAD

**Discussion by marat569** [mpv-player/mpv#14277] [Q&A] [answered]
Pretty much the title -- Is there a way to setup input.conf so when I press a hotkey, MPV will calculate the frame's MaxCLL and output it as text like "MaxCLL 1000"

Also I'm not talking about the MaxCLL in the metadata, but the MaxCLL that is calculated by `hdr-peak-percentile` -- The value you see tonemappers use like in the picture here:
![image](https://github.com/mpv-player/mpv/assets/7017110/1f5f8abf-1b97-4ce6-83cb-a762dc87565b)

So if I was to press said hotkey with that JXL image open, it would read out "MaxCLL 1000"

**Comment by kasper93** [comment, marked as answer]
To show only PQ value, you can use:
```
n show-text ${video-out-params/max-pq-y}
```
To show also converted value in cd/m², you can use:
``` lua
local function pq_eotf(x)
    if not x then
        return x;
    end

    local PQ_M1 = 2610.0 / 4096 * 1.0 / 4
    local PQ_M2 = 2523.0 / 4096 * 128
    local PQ_C1 = 3424.0 / 4096
    local PQ_C2 = 2413.0 / 4096 * 32
    local PQ_C3 = 2392.0 / 4096 * 32

    x = x ^ (1.0 / PQ_M2)
    x = math.max(x - PQ_C1, 0.0) / (PQ_C2 - PQ_C3 * x)
    x = x ^ (1.0 / PQ_M1)
    x = x * 10000.0

    return x
end

mp.add_key_binding('n', function() 
	local pq = mp.get_property_native('video-out-params/max-pq-y')
	if not pq then
		return
	end
	mp.commandv('show-text', string.format("%.2f cd/m² (%.2f%% PQ)", pq_eotf(pq), pq * 100))
end)

```

**Reply by marat569** [reply]
So I used the script, and it works for videos, but sadly not .avif or .jxr images -- and thats where I'd like to get the most value out of the command

I'll attach 2 random hdr screenshots 

HDR Screenshots do not usually have traditional HDR metadata, but the MaxCLL can be calculated if target-peak and tone mapping is active, as you can see in the screenshot above 

I don't want to tonemap HDR screenshots, just press a button and get the MaxCLL values that the tonemappper would use (the 1000 thats underlined in the image in the original post)

[screenies.zip](https://github.com/user-attachments/files/15526559/screenies.zip)

Edit: If I force tonemapping on, the .lua script works in HDR Images -- but ideally I'd like it to work with tonemapping off :c 

Edit 2: Sometimes the script works in the images, sometimes it throws a lua error saying pq doesn't exist (same image)

**Reply by kasper93** [reply]
> Edit: If I force tonemapping on, the .lua script works in HDR Images -- but ideally I'd like it to work with tonemapping off :c

You mentioned peak percentile, so I assumed you want to get peak detection value. If tonemapping is disabled it won't be available. You can use [video-out-params/max-luma](https://mpv.io/manual/master/#command-interface-video-params/max-luma) or [video-out-params/max-cll](https://mpv.io/manual/master/#command-interface-video-params/max-cll) if you aim for static metadata. Also if `video-out-params` is not available, you can use `video-params` which should be always there..

> Edit 2: Sometimes the script works in the images, sometimes it throws a lua error saying pq doesn't exist (same image)

I've added a check for that, but you should extend the script for your needs, it was only an example.


EDIT:

Note that all those values are in stats.lua (shift+i) page 1 if you want to look at them.

**Reply by marat569** [reply]
> > Edit: If I force tonemapping on, the .lua script works in HDR Images -- but ideally I'd like it to work with tonemapping off :c
> 
> You mentioned peak percentile, so I assumed you want to get peak detection value. If tonemapping is disabled it won't be available. You can use [video-out-params/max-luma](https://mpv.io/manual/master/#command-interface-video-params/max-luma) or [video-out-params/max-cll](https://mpv.io/manual/master/#command-interface-video-params/max-cll) if you aim for static metadata. Also if `video-out-params` is not available, you can use `video-params` which should be always there..
> 
> > Edit 2: Sometimes the script works in the images, sometimes it throws a lua error saying pq doesn't exist (same image)
> 
> I've added a check for that, but you should extend the script for your needs, it was only an example.
> 
> EDIT:
> 
> Note that all those values are in stats.lua (shift+i) page 1 if you want to look at them.

The reason I made this issue is because they're not on the shift + i page for certain image formats (HDR PNG, JXL) and sometimes wrong for HDR .AVIF

The only way to get an accurate number is `hdr-peak-percentile=1000` and engage tonemapping via `target-peak` and see what the tonemapper is trying to do

I've attached 2 images in the last post that are a good example

**Reply by kasper93** [reply]
> The only way to get an accurate number is hdr-peak-percentile=1000 and engage tonemapping via target-peak and see what the tonemapper is trying to do

Ok, so you can do it with `video-out-params/max-pq-y`, no other way.

**Reply by marat569** [reply]
> > The only way to get an accurate number is hdr-peak-percentile=1000 and engage tonemapping via target-peak and see what the tonemapper is trying to do
> 
> Ok, so you can do it with `video-out-params/max-pq-y`, no other way.

`video-out-params/max-pq-y` works and accurately reports the MaxCLL  (your script) 

But here is the issue

So to start off, in my conf HDR images do not have a target peak

If I just open the image and press n (your script) -- nothing happens

If I open the image, and press h; which engages the profile below, and then press n (your script) -- it will accurately display the MaxCLL
```ini
[forceTonemapping]
profile-desc="forceTonemapping"
target-peak=1000
tone-mapping=bt.2446a
contrast=0 # Reset contrast value
```

The question I have is if I can get it to calculate and display the MaxCLL without having to engage `target-peak` / "press h"

**Reply by kasper93** [reply]
> The question I have is if I can get it to calculate and display the MaxCLL without having to engage target-peak / "press h"

Not with mpv. If tone-mapping is not needed / used it won't be calculated.

**Reply by marat569** [reply]
> > The question I have is if I can get it to calculate and display the MaxCLL without having to engage target-peak / "press h"
> 
> Not with mpv. If tone-mapping is not needed / used it won't be calculated.

Thanks for the help anyway!

Whats interesting is I figured out a workaround -- `target-peak=4000`

Even if tonemapping is not engaged, as long as target-peak is a thing in the conf, the script will work -- so this auto-profile works for HDR Images

Thanks a lot man!

```ini
[hdrImages]
#profile-cond=get('filename', ''):match('%.avif$') ~= nil
profile-cond=(p.video_params.gamma == "pq" or p.video_params.gamma == "hlg") and p['current-tracks/video/image'] == true
contrast=0 #reset contrast
hdr-peak-percentile=100 #Testing more accurate MaxCLL calculations
hwdec=no #avoids hardware decoding errors
deband=no
keep-open=yes #So MPV stays open kek
target-peak=4000 #For MaxCLL Calculation
```
